src/client/canvas/svg/svg.tsx

```diff
@@ -0,0 +1,82 @@
+import { Component } from 'react'
+import * as React from 'react'
+
+import { Transform } from '../../math/transform'
+import { Appearance } from '../appearance/appearance'
+import { BasicAppearance } from '../appearance/basic_appearance'
+import { LineAppearance } from '../appearance/line_appearance'
+import { ArcGeometry } from '../geometry/arc_geometry'
+import { ArrowGeometry } from '../geometry/arrow_geometry'
+import { CircleGeometry } from '../geometry/circle_geometry'
+import { LineGeometry } from '../geometry/line_geometry'
+import { MarkerGeometry } from '../geometry/marker_geometry'
+import { PolygonGeometry } from '../geometry/polygon_geometry'
+import { TextGeometry } from '../geometry/text_geometry'
+import { Group as GroupGeometry } from '../object/group'
+import { Object2d } from '../object/object2d'
+import { Geometry } from '../object/shape'
+import { Shape } from '../object/shape'
+
+import { Arc } from './arc'
+import { Arrow } from './arrow'
+import { Circle } from './circle'
+import { Group } from './group'
+import { Line } from './line'
+import { Marker } from './marker'
+import { Polygon } from './polygon'
+import { Text } from './text'
+
+export function appearance(appearance: Appearance): {} {
+  if (appearance instanceof BasicAppearance) {
+    return {
+      fill: appearance.fillStyle,
+      strokeWidth: appearance.lineWidth,
+      stroke: appearance.strokeStyle,
+    }
+  } else if (appearance instanceof LineAppearance) {
+    return {
+      strokeLinecap: appearance.lineCap,
+      strokeLinejoin: appearance.lineJoin,
+      strokeDashoffset: appearance.lineDashOffset,
+      strokeWidth: appearance.lineWidth,
+      stroke: appearance.strokeStyle,
+    }
+  } else {
+    throw new Error(`Unsupported appearance type ${appearance}`)
+  }
+}
+
+export function transform(transform: Transform): string {
+  const s = transform.scale
+  const r = (180.0 / Math.PI) * transform.rotate // SVG rotations are in degrees
+  const t = transform.translate
+  return `translate(${t.x}, ${t.y}) rotate(${r}) scale(${s.x}, ${s.y})`
+}
+
+export function viewForGeometry(obj: Object2d, index: number, world: Transform): JSX.Element {
+
+  // TODO wrap things in appearance tags
+  if (obj instanceof GroupGeometry) {
+    return <Group key={index} model={obj} world={world} />
```

Comment:
Having the `key` prop set here forces that these components can only be rendered within a list.

Moving the responsibility for setting the `key` prop to whoever is rendering the list would be ideal here.

In order to achieve that, the return type of this function must be a renderable `ComponentType`, not a `JSX.Element`.

I'll have a spike and post the gist back here.

Reply:
Have a look at [this gist](https://gist.github.com/BrendanAnnable/0d8e2539dbdcc617dbd4f12aeeae2c1d).